Compiler: use Colorize.on_tty_only!

MakeNowJust opened the pull request:
`Colorize.on_tty_only!` is introduced in #4271, but it is not used in compiler. I think it is useful that colorize is turned on/off depending on tty detection result.

In addition, I fixed `bin/crystal` wrapper script to detect TTY. `bin/crystal` script's `SUPPORTS_COLORS` is really buggy, at least it isn't working correct now.

RX14 commented:
Could you seperate the changes to bin/crystal and the compiler into seperate commits? The core team might even want them in seperate prs.

MakeNowJust commented:
ok. Please wait.

MakeNowJust commented:
@RX14 I can separate this PR into the wrapper bug fix and others (using `Colorize.on_tty_support!` in compiler). This bug fix is just one line, however it causes a conflict with this PR surely (because this PR added TTY detection into this line). Do you want it really?

RX14 commented:
@MakeNowJust i'm not sure I understand what the conflict is.

MakeNowJust commented:
@RX14 Bug fix patch:

```diff
diff --git a/bin/crystal b/bin/crystal
index 54704199d..8b4380f30 100755
--- a/bin/crystal
+++ b/bin/crystal
@@ -115,7 +115,7 @@ __has_colors() {
     return 1
   fi
 }
-SUPPORTS_COLORS=$(__has_colors)
+SUPPORTS_COLORS=$(__has_colors && echo true || echo false)
 __error_msg() {
   if $SUPPORTS_COLORS; then
     # bold red coloring
```

MakeNowJust commented:
I didn't intend to fix a bug. I implemented TTY detection on the wrapper script correctly, then I fixed a bug unexpectedly.

Why cause a problem when I make a PR about `Colorize`!!!!

MakeNowJust commented:
This PR diff is a little large, but it is really simple fix. I will explain:

  - Added `Colorize.on_tty_only!` to `Crystal::Command#run`.
  - Fixed `--no-color` option to `Colorize.enabled = false`, and added `--color` option as `Colorize.enabled = true`.
  - Removed `@color` instance variable and `colorize`, `with_color` method, then replaced `colorize(obj)` to `obj.colorize`. In addition, removed `.toggle(@color)` on `commands/formatter.cr`.
  - Two specs about compiler is broken, fixed it.
  - The wrapper is fixed to use colorize on TTY only.

That's all.

straight-shoota commented:
@MakeNowJust could you rebase this PR on master?

MakeNowJust commented:
@straight-shoota rebased.

straight-shoota commented on spec/compiler/semantic/generic_class_spec.cr:
```diff
@@ -636,7 +636,9 @@ describe "Semantic: generic class" do
     begin
       semantic(nodes)
     rescue ex : TypeException
+      old_enabled, Colorize.enabled = Colorize.enabled?, false
```
This would be better as two statements.

MakeNowJust replied:
Thank you but I don't think that.

straight-shoota commented:
I'd suggest to add the option to explicitly set color mode to `auto` (which is the default). This would be additional to the already included `--color` and `--no-color`. 
Example from `man git-diff(1)`:
```
--color[=<when>]
Show colored diff.  --color (i.e. without =<when>) is the same as --color=always.  <when> can be one of always, never, or auto. [...]
 --no-color
Turn off colored diff. This can be used to override configuration settings. It is the same as --color=never.
```

I think this would be the best way to handle this.

MakeNowJust commented:
@straight-shoota It was implemented in #3925...

Sija commented on src/compiler/crystal/command.cr:
```diff
@@ -525,9 +519,39 @@ class Crystal::Command
+      case color
+      when "auto"
+        Colorize.on_tty_only!
+      when "always"
```
`color = "always" if color.empty?` => `when "always", .empty?`

MakeNowJust replied:
`when "always", ""`

MakeNowJust replied:
Personally, `color = "always" if color.empty?` shows `"always"` as default clearly, so I would keep this.

straight-shoota replied:
Maybe, but it adds an completely unnecessary comparison and assignment.

straight-shoota replied:
It should be all in one `case` statement, not a `case` plus an `if`.

MakeNowJust commented:
`bin/crystal run --color a.cr` means `bin/crystal run --color=a.cr`, so `--color`'s default value does not make a sense. I'll remove this.

straight-shoota commented:
Why not use `--color=[always,never,auto]`?

MakeNowJust commented:
@straight-shoota It does not change meaning.

straight-shoota commented:
Yes it does. If `--color` expects a `=` to be assigned a value, `--color a.cr` will not be confused.

MakeNowJust commented:
Yes but it is not implemented.

straight-shoota commented:
You mean `OptionParser` doesn't differentiate between `--color=<when>` and `--color <when>`? Then maybe it should or you could implement this differently.

The commands should provide an arg-less `--color` flag to enable colorization.

bew commented:
Maybe it has already been said, but to me, the `<when>` is useless:
- `--color` argument: force colors
- `--no-color` argument: disable colors
- no color-related argument: automatic colors (on tty)

straight-shoota commented:
It is not useless but adds expressiveness.

It makes it easy to inject a configuration value into the command line arguments, like `crystal --color=$MY_COLOR_SETTING ...` instead of some complicated decision tree between `--color`, `--no-color` and ``.

larubujo commented:
more complex things. global state in colorize used by all apps. global state is bad. the design is being ruined. more command line switches. the direction should be to simplify things.

straight-shoota commented:
@larubujo This PR is about the compiler, not "all apps".

MakeNowJust commented:
> The commands should provide an arg-less --color flag to enable colorization.

I totally agree with this. But it can be merged this before implementing this feature in `OptionParser`, and then fixes this in PR about `OptionParser`.

RX14 commented:
I personally prefer @bew's suggestion.

MakeNowJust commented:
@straight-shoota I want to revert `--color=when` support for now because it blocks to merge this PR and I don't think this option is needed strongly. If you want this feature really, then you can create a new pull request after fixed `OptionParser`.

MakeNowJust commented:
Sorry... 🙇

wontruefree commented:
@MakeNowJust are you still working on this?
